cpp/tests/hashing/sha1_test.cpp

```diff
@@ -0,0 +1,204 @@
+     "cb73203438ab46ea54491c53e288a2703c440c4a",
+     "c595ebd13a785c1c2659e010a42e2ff9987ef51f",
+     "4ffaf61804c55b8c2171be548bef2e1d0baca17a",
+     "595965dd18f38087186162c788485fe249242131",
+     "a62ca720fbab830c8890044eacbeac216f1ca2e4",
+     "11e16c52273b5669a41d17ec7c187475193f88b3"});
+
+  cudf::test::strings_column_wrapper const sha1_string_results2(
+    {"da39a3ee5e6b4b0d3255bfef95601890afd80709",
+     "fb96549631c835eb239cd614cc6b5cb7d295121a",
+     "e3977ee0ea7f238134ec93c79988fa84b7c5d79e",
+     "f6f75b6fa3c3d8d86b44fcb2c98c9ad4b37dcdd0",
+     "c7abd431a775c604edf41a62f7f215e7258dc16a",
+     "153fdf20d2bd8ae76241197314d6e0be7fe10f50",
+     "8c3656f7cb37898f9296c1965000d6da13fed64e",
+     "b4a848399375ec842c2cb445d98b5f80a4dce94f"});
+
+  using limits = std::numeric_limits<int32_t>;
+  cudf::test::fixed_width_column_wrapper<int32_t> const ints_col(
+    {0, 100, -100, limits::min(), limits::max(), 1, 2, 3});
+
+  // Different truth values should be equal
+  cudf::test::fixed_width_column_wrapper<bool> const bools_col1({0, 1, 1, 1, 0, 1, 1, 1});
+  cudf::test::fixed_width_column_wrapper<bool> const bools_col2({0, 1, 2, 255, 0, 1, 2, 255});
+
+  auto const string_input1       = cudf::table_view({strings_col});
+  auto const string_input2       = cudf::table_view({strings_col, strings_col});
+  auto const sha1_string_output1 = cudf::hashing::sha1(string_input1);
+  auto const sha1_string_output2 = cudf::hashing::sha1(string_input2);
+  EXPECT_EQ(string_input1.num_rows(), sha1_string_output1->size());
+  EXPECT_EQ(string_input2.num_rows(), sha1_string_output2->size());
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(sha1_string_output1->view(), sha1_string_results1);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(sha1_string_output2->view(), sha1_string_results2);
+
+  auto const input1       = cudf::table_view({strings_col, ints_col, bools_col1});
+  auto const input2       = cudf::table_view({strings_col, ints_col, bools_col2});
+  auto const sha1_output1 = cudf::hashing::sha1(input1);
+  auto const sha1_output2 = cudf::hashing::sha1(input2);
+  EXPECT_EQ(input1.num_rows(), sha1_output1->size());
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(sha1_output1->view(), sha1_output2->view());
+}
+
+TEST_F(SHA1HashTest, MultiValueNulls)
+{
+  // Nulls with different values should be equal
+  cudf::test::strings_column_wrapper const strings_col1(
+    {"",
+     "Different but null!",
+     "A very long (greater than 128 bytes/char string) to execute a multi hash-step data point in "
+     "the hash function being tested. This string needed to be longer.",
+     "All work and no play makes Jack a dull boy",
+     "!\"#$%&\'()*+,-./0123456789:;<=>?@[\\]^_`{|}~"},
+    {1, 0, 0, 1, 0});
+  cudf::test::strings_column_wrapper const strings_col2(
+    {"",
+     "Another string that is null.",
+     "Very different... but null",
+     "All work and no play makes Jack a dull boy",
+     ""},
+    {1, 0, 0, 1, 1});  // empty string is equivalent to null
```

Comment:
I don't we should be testing with unsanitized nulls.

Reply:
Agreed. I am fixing this for SHA hashing code, and backporting that fix to the MD5 code. The MD5 changes are done in 01881154feacfbc1fca4ee0e3b0a40ef56b5802f and I'm working on SHA updates now.

Reply:
This will need a more thorough follow-up for other hashing algorithms (MD5 and SHA are easy to fix together because of their similarities). I have noted that in https://github.com/rapidsai/cudf/issues/13706#issuecomment-1901194560.

Reply:
The SHA removal of unsanitized null tests is done in 186eace705f00b815c2a8962a8a92fbdda922adf and 65358ff6f297193e90ccd8f849da5cd8d482e84a. Resolving.